### PR TITLE
andreer/endpoint certificate maintainer

### PR DESCRIPTION
**Pull request by andreer**

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    private void deployRefreshedCertificates() {
+        var now = clock.instant();
+        curator.readAllEndpointCertificateMetadata().forEach((applicationId, endpointCertificateMetadata) ->
+                endpointCertificateMetadata.lastRefreshed().ifPresent(lastRefreshTime -> {
+                    Instant refreshTime = Instant.ofEpochSecond(lastRefreshTime);
+                    if (now.isAfter(refreshTime.plus(7, ChronoUnit.DAYS))) {
+
+                        controller().jobController().jobs(applicationId).forEach(job ->
+                                controller().jobController().jobStatus(new JobId(applicationId, JobType.fromJobName(job.jobName()))).lastTriggered().ifPresent(run -> {
+                                    if (run.start().isBefore(refreshTime) && job.isProduction()) {
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java:108]:
```suggestion
                                    if (run.start().isBefore(refreshTime) && job.isProduction() && job.isDeployment()) {
```
lest you trigger production tests needlessly.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainerTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        maintainer.maintain();
+
+        tester.clock().advance(Duration.ofDays(8));
+
+        deploymentContext.assertNotRunning(productionUsWest1);
+
+        maintainer.maintain();
+
+        deploymentContext.assertRunning(productionUsWest1);
+    }
```
Comment (jonmv) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainerTest.java:125]: 👍

**Review by jonmv — COMMENTED** [review]
Integration with jobs LGTM :)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            if (latestAvailableVersion.isPresent() && latestAvailableVersion.getAsInt() > endpointCertificateMetadata.version()) {
+                var refreshedCertificateMetadata = endpointCertificateMetadata.withVersion(latestAvailableVersion.getAsInt());
+                curator.writeEndpointCertificateMetadata(applicationId, refreshedCertificateMetadata);
+                // TODO andreer: Certificate not validated here! Is that OK?
+                deploymentTrigger.triggerChange(applicationId, Change.empty()); // TODO andreer: No idea if this does what I want, or anything at all
+            }
+        }));
+    }
+
+    private OptionalInt latestVersionInSecretStore(EndpointCertificateMetadata originalCertificateMetadata) {
```
Comment (tokle) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java:118]: Do we somehow want to catch (report/alert) situations where certversion<keyversion for a long time?

**Review by tokle — APPROVED** [review]

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java (hunk 3 above)
Comment (andreer) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/EndpointCertificateMaintainer.java:118]: It should cause an alert to oncall after a week, as the cert nears expiry. I think that's good enough

**Pull request merged**